src/client/debugger/jupyter/debuggingManager.ts

```diff
@@ -218,7 +218,15 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
             return;
         }
 
-        if (this.notebookInProgress.has(editor.document) || this.isDebugging(editor.document)) {
+        if (this.notebookInProgress.has(editor.document)) {
+            return;
+        }
+
+        if (this.isDebugging(editor.document)) {
+            this.updateToolbar(true);
```

Comment:
Why do you have to do this here? If you are already debugging, shouldn't it already be in this state?

Reply:
it should, but that's the buggy state you get trapped in.

this would correct the context keys

Reply:
Not sure i like this, feels like we're re-executing some code & that seems to fix a problem (but we don't know why).
I would like us to fix the root cause, why is this required in the first place

Reply:
The root cause is out of our hands, its debugpy freezing inside ipykernel.

And I disagree, we should be trying to fix root causes, but since the root cause is not on our code, we should also try to fix it if it happens.

Reply:
> its debugpy freezing inside ipykernel.

If that's the cause, how can updating the toolbar fix anything in the kernel (i.e. updating a UI will not address anything in ipykernel freezing).

Reply:
It seems like the root cause is us having the context keys out of sync with the RBL lifecycle. Your latest commit is a good one towards fixing that, I think previously we reset the context keys too early, before debugging actually stopped.

Fundamentally, at this point in the code, two things are out of sync. It's hard to point at one and say this one is always right. It should be possible to write the code so that they are always in sync.